src/System/Information.hs

```diff
@@ -54,26 +60,39 @@ instance Show OS where
 -- | Get the current OS' name
 getOS :: IO (Maybe OS)
 getOS = do
-  eResult <- try $ readProcess
+  eResult <- try $
 #ifdef darwin_HOST_OS
-    "sw_vers" [] ""
+    readProcess "sw_vers" [] ""
 #elif linux_HOST_OS
-    "lsb_release" ["-d"] ""
+    readProcess "lsb_release" ["-d"] ""
 #elif mingw32_HOST_OS
-    "systeminfo" [] ""
+    getWindowsOS
 #endif
 
   case eResult of
     Left (_ :: SomeException) ->
       either (const Nothing :: SomeException -> Maybe OS)
              (Just . OS) <$> try (readProcess "tr" ["-d", "\\n"] =<< readProcess "uname" ["-sr"] "")
-    Right res -> pure $ OS <$> flip parseLineAfter res
+    Right res -> pure $ OS <$>
 #ifdef darwin_HOST_OS
-      "ProductName:"
+      flip parseLineAfter res "ProductName:"
 #elif linux_HOST_OS
-      "Description:"
+      flip parseLineAfter res "Description:"
 #elif mingw32_HOST_OS
-      "OS Name:"
+      Just res
+#endif
+
+#ifdef mingw32_HOST_OS
+getWindowsOS :: IO String
+getWindowsOS = alloca $ \hres -> do
+  os_str <- c_getOS hres
+  res <- peek hres :: IO CInt
+  if res == 0
+  then peekCWString os_str
+  else fail ("getOS failed: " ++ show res)
+
+foreign import ccall safe "getOS"
+  c_getOS :: Ptr CInt -> IO CWString
```

Comment:
This could be pure function, because it will not change value during whole program  execution guaranteed.
```haskell
c_getOS :: Ptr CInt -> CWString
```
and in the same way:
```haskell
getWindowsOS :: String
```
But asynchronous exceptions should be masked during execution of it.